### PR TITLE
Bump Golang to 1.23.1

### PR DESCRIPTION
**Pull request by yuntanghsu**
In order to ensure TLS 1.3 is supportted. We bump Golang version to 1.23.1

In this patch, we do:
1. Bump golangci to v1.61.0 in order to support golang v1.23.1
2. To increase flexibility for downstream consumers, we set a lower Go version in `go.mod` for API module (nsx-operator/pkg/apis), allowing them to choose their preferred Go version. However, we will keep the main module (nsx-operator) on the latest Go version. Currently, we set Go 1.21 for the API module.
3. Update controller-runtime to v0.19.0 for this issue:
https://github.com/kubernetes-sigs/controller-runtime/issues/2925
4. Update controller-gen to v16.5 to resolve the error below.
```
/Users/hsuy/Desktop/nsx-operator/bin/controller-gen rbac:roleName=manager-role crd webhook paths="github.com/vmware-tanzu/nsx-operator/pkg/apis/legacy/v1alpha1" output:crd:artifacts:config=build/yaml/crd/legacy/
/Users/hsuy/go/pkg/mod/k8s.io/api@v0.31.2/core/v1/doc.go:21:1: missing argument "" (at <input>)
/Users/hsuy/go/pkg/mod/k8s.io/api@v0.31.2/core/v1/doc.go:21:1: missing argument "" (at <input>)
k8s.io/api/core/v1:-: unknown type "k8s.io/api/core/v1".ConditionStatus
k8s.io/api/core/v1:-: unknown type "k8s.io/api/core/v1".ConditionStatus
github.com/vmware-tanzu/nsx-operator/pkg/apis/legacy/v1alpha1:-: unable to locate schema for type "k8s.io/api/core/v1".ConditionStatus
```

Signed-off-by: Yun-Tang Hsu <yun-tang.hsu@broadcom.com>

### CODE REVIEW

**Review by zhengxiexie — COMMENTED** [review]
Hi, we should ensure `make generated` `make build` succeed, I failed on this branch.

**Comment by TaoZou1** [comment]
I run 
go mod tidy
go clean -cache -modcache -i -r
tazou@C02D44AJMD6R nsx-operator-1 % make build
/Users/tazou/go/src/nsx-operator-1/bin/controller-gen object:headerFile="hack/boilerplate.go.txt" paths="./pkg/apis/..."

===> Formatting Go files <===
go vet ./...
# github.com/vmware-tanzu/nsx-operator/github.com/vmware-tanzu/nsx-operator/pkg/apis/vpc/v1alpha1
vet: github.com/vmware-tanzu/nsx-operator/pkg/apis/vpc/v1alpha1/zz_generated.deepcopy.go:16:11: undefined: AddressBinding

**Comment by yuntanghsu** [comment]
I tried to run `make generated` for previous commit on the main branch. All them failed for 
```
go: github.com/vmware-tanzu/nsx-operator/pkg/client/clientset/versioned/fake imports
	github.com/vmware-tanzu/nsx-operator/pkg/apis/vpc/v1alpha1: module github.com/vmware-tanzu/nsx-operator/pkg/apis@latest found (v1.0.0), but does not contain package github.com/vmware-tanzu/nsx-operator/pkg/apis/vpc/v1alpha1
```
Is it a regression?

**Comment by zhengxiexie** [comment]
> I tried to run `make generated` for previous commits on the main branch. All of them failed for
> 
> ```
> go: github.com/vmware-tanzu/nsx-operator/pkg/client/clientset/versioned/fake imports
> 	github.com/vmware-tanzu/nsx-operator/pkg/apis/vpc/v1alpha1: module github.com/vmware-tanzu/nsx-operator/pkg/apis@latest found (v1.0.0), but does not contain package github.com/vmware-tanzu/nsx-operator/pkg/apis/vpc/v1alpha1
> ```
> 
> Is it a regression?

first `make build`

**Comment by yuntanghsu** [comment]
Looks like we need to run `go env -w GOPROXY='direct'` locally.
After running this command, `make generated`, `make build` pass.
Thanks for @yanjunz97

**Comment by zhengxiexie** [comment]
/Users/zhengxie/Code/nsx-operator/bin/controller-gen rbac:roleName=manager-role crd webhook paths="github.com/vmware-tanzu/nsx-operator/pkg/apis/legacy/v1alpha1" output:crd:artifacts:config=build/yaml/crd/legacy/
/Users/zhengxie/.asdf/installs/golang/1.23.1/packages/pkg/mod/k8s.io/api@v0.31.2/core/v1/doc.go:21:1: missing argument "" (at <input>)
/Users/zhengxie/.asdf/installs/golang/1.23.1/packages/pkg/mod/k8s.io/api@v0.31.2/core/v1/doc.go:21:1: missing argument "" (at <input>)
k8s.io/api/core/v1:-: unknown type "k8s.io/api/core/v1".ConditionStatus
k8s.io/api/core/v1:-: unknown type "k8s.io/api/core/v1".ConditionStatus
github.com/vmware-tanzu/nsx-operator/pkg/apis/legacy/v1alpha1:-: unable to locate schema for type "k8s.io/api/core/v1".ConditionStatus
Error: not all generators ran successfully
run `controller-gen rbac:roleName=manager-role crd webhook paths=github.com/vmware-tanzu/nsx-operator/pkg/apis/legacy/v1alpha1 output:crd:artifacts:config=build/yaml/crd/legacy/ -w` to see all available markers, or `controller-gen rbac:roleName=manager-role crd webhook paths=github.com/vmware-tanzu/nsx-operator/pkg/apis/legacy/v1alpha1 output:crd:artifacts:config=build/yaml/crd/legacy/ -h` for usage
make: *** [manifests] Error 1

go version go1.23.1 darwin/arm64

**Review by TaoZou1 — DISMISSED** [review]

**Review by zhengxiexie — DISMISSED** [review]

File: build/yaml/crd/vpc/crd.nsx.vmware.com_securitypolicies.yaml [hunk 1]
```diff
@@ -448,7 +448,6 @@ spec:
                             description: Port is the name or port number.
                             x-kubernetes-int-or-string: true
                           protocol:
-                            default: TCP
```
Comment (timdengyun) [inline build/yaml/crd/vpc/crd.nsx.vmware.com_securitypolicies.yaml:451]: why we have this CRD change?

Comment (yuntanghsu) [reply build/yaml/crd/vpc/crd.nsx.vmware.com_securitypolicies.yaml:451]: They are generated after I upgrade controller-gen to v.0.16.5. Is it an expected behavior? Or we shouldn't let controller-gen to modify the crd?

File: build/yaml/crd/legacy/nsx.vmware.com_securitypolicies.yaml [hunk 2]
```diff
@@ -448,7 +448,6 @@ spec:
                             description: Port is the name or port number.
                             x-kubernetes-int-or-string: true
                           protocol:
-                            default: TCP
```
Comment (timdengyun) [inline build/yaml/crd/legacy/nsx.vmware.com_securitypolicies.yaml:451]: Thanks @yuntanghsu,
For this T1 CRD, 
could you help have a try to change CRD definition:
https://github.com/vmware-tanzu/nsx-operator/blob/main/pkg/apis/legacy/v1alpha1/securitypolicy_types.go#L103
 
to add one kubebuilder directive:
`// +kubebuilder:default=TCP`

to see if the controller-gen can generate default value for protocol field.

Comment (yuntanghsu) [reply build/yaml/crd/legacy/nsx.vmware.com_securitypolicies.yaml:451]: Thanks @timdengyun , I've added back. Please take a look again. Thanks!

Comment (timdengyun) [reply build/yaml/crd/legacy/nsx.vmware.com_securitypolicies.yaml:451]: Hi @yuntanghsu I see you changed VPC securitypolicy_types.go, it looks like working. 
Could you also help to change legacy T1 type:
https://github.com/vmware-tanzu/nsx-operator/blob/main/pkg/apis/legacy/v1alpha1/securitypolicy_types.go#L103
and re generated CRD yaml for T1. Thanks

**Review by timdengyun — APPROVED** [review]
LGTM

**Review by zhengxiexie — APPROVED** [review]
LGTM

**Pull request merged**